bluesky/abc.py

```diff
@@ -0,0 +1,264 @@
+        a suggestion to enable automated tools to provide helpful information
+        with minimal guidance from the user. See :ref:`hints`.
+        """
+        ...
+
+    def trigger(self) -> Status:
+        """Return a ``Status`` that is marked done when the device is done triggering.
+
+        If the device does not need to be triggered, simply return a ``Status``
+        that is marked done immediately.
+        """
+        ...
+
+    def read(self) -> Configuration:
+        """Return an OrderedDict mapping field name(s) to values and timestamps.
+
+        The field names must be strings. The values can be any JSON-encodable
+        type or a numpy array, which the RunEngine will convert to (nested)
+        lists. The timestamps should be UNIX time (seconds since 1970).
+
+        Example return value:
+
+        .. code-block:: python
+
+            OrderedDict(('channel1',
+                         {'value': 5, 'timestamp': 1472493713.271991}),
+                         ('channel2',
+                         {'value': 16, 'timestamp': 1472493713.539238}))
+        """
+        ...
+
+    def describe(self) -> Configuration:
+        """Return an OrderedDict with exactly the same keys as the ``read``
+        method, here mapped to metadata about each field.
+
+        Example return value:
+
+        .. code-block:: python
+
+            OrderedDict(('channel1',
+                         {'source': 'XF23-ID:SOME_PV_NAME',
+                          'dtype': 'number',
+                          'shape': []}),
+                        ('channel2',
+                         {'source': 'XF23-ID:SOME_PV_NAME',
+                          'dtype': 'number',
+                          'shape': []}))
+
+        We refer to each entry as a "data key." These fields are required:
+
+        * source (a descriptive string --- e.g., an EPICS Process Variable)
+        * dtype: one of the JSON data types: {'number', 'string', 'array'}
+        * shape: list of integers (dimension sizes) --- e.g., ``[5, 5]`` for a
+          5x5 array. Use empty list ``[]`` to indicate a scalar.
+
+        Optional additional fields (precision, units, etc.) are allowed.
+        The optional field ``external`` should be used to provide information
+        about references to externally-stored data, such as large image arrays.
+        """
+        ...
+
+    def read_configuration(self) -> Configuration:
+        """Same API as ``read`` but for slow-changing fields related to configuration.
+
+        (e.g., exposure time)
+
+        These will typically be read only once per run.
+
+        Of course, for simple cases, you can effectively omit this complexity
+        by returning an empty dictionary.
+        """
+        ...
+
+    def describe_configuration(self) -> Configuration:
+        """Same API as ``describe``, but corresponding to the keys in ``read_configuration``."""
+        ...
+
+
+@runtime_checkable
+class Configurable(Protocol):
```

Comment:
I've used `configure`: https://github.com/thomascobb/bluefly/blob/0ee7ea837d5f74e3cdc69e2315e9beed28dd7766/bluefly/core.py#L411
but I thought it belonged with `read_configuration` and `describe_configuration`. I could be wrong though...

Reply:
I was wondering about that. from what I can tell read/describe configuration are required for _all_ valid devices, but `configure` is only required for devices which can _update_ their configuration (static config is available on all devices)

Reply:
@tacaswell and I now consider `configure` to be a misfeature. It is similar to `set` (or a bunch of `set`s) with the side-effect of notifying the RunEngine, "Any Event Descriptor documents you emitted with these configuration readers are now out of date, so if/when you emit new readings from this device, emit a fresh Event Descriptor as well."

The major problem with this is that the configuration setting could be changed _externally_. (Example: Someone runs `caput` or pokes a button in CSStudio.) In that way, the Event Descriptor would become "wrong" but the RunEngine would never know. The proper way to handle this is to let ophyd _notify_ the RunEngine that some signal that was dubbed "configuration" has changed under it, and that it should reissue an Event Descriptor.

We have known about this problem for awhile (maybe as far back as 2016) but it has never been a priority because `configure` is extremely rarely used. Let's take this moment to formally deprecate it, and drop this class entirely from the PR.

In fact, because it is so rarely used, we discussed going as far as immediately repurposing the method name along the lines proposed by @thomascobb elsewhere in the conversation on this PR.